src/library/scala/collection/immutable/RedBlackTree.scala

```diff
@@ -372,42 +510,239 @@ private[collection] object RedBlackTree {
    *
    * An alternative is to implement the these classes using plain old Java code...
    */
-  sealed abstract class Tree[A, +B](
-                                     @(`inline` @getter) final val key: A,
-                                     @(`inline` @getter) final val value: B,
-                                     @(`inline` @getter) final val left: Tree[A, B],
-                                     @(`inline` @getter) final val right: Tree[A, B]) extends Serializable
+  final class Tree[A, +B](
+                           @(inline @getter @setter)     private var _key: A,
+                           @(inline @getter @setter)     private var _value: AnyRef,
+                           @(inline @getter @setter)     private var _left: Tree[A, _],
+                           @(inline @getter @setter)     private var _right: Tree[A, _],
+                           @(inline @getter @setter)     private var _count: Int)
   {
-    @(`inline` @getter) final val count: Int = 1 + RedBlackTree.count(left) + RedBlackTree.count(right)
-    def black: Tree[A, B]
-    def red: Tree[A, B]
-  }
-  final class RedTree[A, +B](key: A,
-                             value: B,
-                             left: Tree[A, B],
-                             right: Tree[A, B]) extends Tree[A, B](key, value, left, right) {
-    override def black: Tree[A, B] = BlackTree(key, value, left, right)
-    override def red: Tree[A, B] = this
-    override def toString: String = "RedTree(" + key + ", " + value + ", " + left + ", " + right + ")"
-  }
-  final class BlackTree[A, +B](key: A,
-                               value: B,
-                               left: Tree[A, B],
-                               right: Tree[A, B]) extends Tree[A, B](key, value, left, right) {
-    override def black: Tree[A, B] = this
-    override def red: Tree[A, B] = RedTree(key, value, left, right)
-    override def toString: String = "BlackTree(" + key + ", " + value + ", " + left + ", " + right + ")"
+    def isMutable: Boolean = (_count & 0x7FFFFFFF) == 0
+    // read only APIs
+    @`inline` final def count = {
+      //devTimeAssert((_count & 0x7FFFFFFF) != 0)
+      _count & 0x7FFFFFFF
+    }
+    //inlined here to avoid outer object null checks
+    @`inline` final def sizeOf(tree:Tree[_,_]) = if (tree eq null) 0 else tree.count
+    @`inline` final def key = _key
+    @`inline` final def value = _value.asInstanceOf[B]
+    @`inline` final def left = _left.asInstanceOf[Tree[A, B]]
+    @`inline` final def right = _right.asInstanceOf[Tree[A, B]]
+    @`inline` final def isBlack = _count < 0
+    @`inline` final def isRed = _count >= 0
+
+    override def toString: String = s"${if(isRed) "RedTree" else "BlackTree"}($key, $value, $left, $right)"
+
+    @`inline` private def initialCount = _count & 0x80000000
+    @`inline` def colourBit = 0x80000000
```

Comment:
At least `initialBlackCount` and `initialRedCount` are duplcated with same-named members of `object RedBlackTree`. Delete these ones?

Reply:
they we duplicated here to avoid the null check on the outer object (as commented 10 lines above). Maybe this is an Intellij issue

Looking at ASM from intellij, with the change we get
I have not looked at the optimised output from scalac though. Is there a way to configure Intellij/ASM to use the optimiser?

I will see if I can have a look at the optimised output later

```
   LINENUMBER 568 L4
    ALOAD 0
    GETSTATIC scala/collection/immutable/NewRedBlackTree$.MODULE$ : Lscala/collection/immutable/NewRedBlackTree$;
    INVOKEVIRTUAL scala/collection/immutable/NewRedBlackTree$.initialBlackCount ()I
    INVOKESPECIAL scala/collection/immutable/NewRedBlackTree$Tree.scala$collection$immutable$NewRedBlackTree$Tree$$_count_$eq (I)V
```
with the existing just the outer reference we get
```
 LINENUMBER 570 L4
    ALOAD 0
    ALOAD 0
    INVOKEVIRTUAL scala/collection/immutable/NewRedBlackTree$Tree.initialBlackCount ()I
    INVOKESPECIAL scala/collection/immutable/NewRedBlackTree$Tree.scala$collection$immutable$NewRedBlackTree$Tree$$_count_$eq (I)V
  
```

I have reduced the visibility of the methods as best I can. (must get scalaclean working  😢 )

Reply:
after some experimentation and discussion updated with https://github.com/scala/scala/pull/8794/commits/1cfc95273981cdc78017a8eec6c8acfded302376 

making a final val in the enclosing objects uses the constant values directly. Also reworked the dependency between the constants and added comments explaining why.